nexus/db-queries/src/db/queries/disk.rs

```diff
@@ -0,0 +1,168 @@
+///
+/// The expected form of the statement is
+///
+/// ```sql
+/// SET attach_instance_id = instance_id,
+///     disk_state = 'attached',
+///     slot = (SELECT 0 + shift AS slot FROM
+///             (SELECT generate_series(0, 8) AS shift
+///              UNION ALL
+///              SELECT generate_series(0, -1) AS shift)
+///             LEFT OUTER JOIN disk
+///             ON (attach_instance_id, slot, time_deleted IS NULL) =
+///                (instance_id, 0 + shift, TRUE)
+///             WHERE slot IS NULL
+///             LIMIT 1)
+/// ```
+///
+/// This fragment can be passed to an `attach_resource` operation by supplying
+/// it as the argument to a `set`, e.g.
+/// `diesel::update(disk::dsl::disk).set(DiskSetClauseForAttach::new(instance_id))`.
+#[derive(Debug, Clone)]
+pub struct DiskSetClauseForAttach {
+    attach_instance_id: Uuid,
+    next_slot: NextDiskSlot,
+}
+
+impl DiskSetClauseForAttach {
+    pub fn new(instance_id: Uuid) -> Self {
+        Self {
+            attach_instance_id: instance_id,
+            next_slot: NextDiskSlot::new(instance_id),
+        }
+    }
+}
+
+impl QueryId for DiskSetClauseForAttach {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl QueryFragment<Pg> for DiskSetClauseForAttach {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        let attached_label =
+            omicron_common::api::external::DiskState::Attached(
+                self.attach_instance_id,
+            )
+            .label();
+
+        // Ideally this code would not have to handcraft the entire SET
+        // statement. It would be cleaner to write something like
+        //
+        // diesel::update(disk::dsl::disk).set(
+        //  (disk::dsl::disk::attach_instance_id.eq(attach_instance_id),
+        //   disk::dsl::disk::disk_state.eq(attached_label),
+        //   disk::dsl::disk::slot.eq(next_slot_query))
+        //
+        // where `next_slot_query` is a wrapper around the `NextItem` query
+        // above that yields the slot-choosing SELECT statement. Diesel provides
+        // a `single_value` adapter function that's supposed to do this: given
+        // a query that returns a single column, it appends a LIMIT 1 to the
+        // query and parenthesizes it so that it can be used as the argument to
+        // a SET statement. The sticky bit is the trait bounds: SingleValueDsl
+        // is implemented for SelectQuery + LimitDsl, and while it's easy enough
+        // to impl SelectQuery for NextDiskSlot, implementing LimitDsl is harder
+        // (and seems to be discouraged by the trait docs)--it is natively
+        // implemented for diesel::Table, but here there's no table type, in
+        // part because the left side of the join is synthesized from thin air
+        // by the NextItem query fragment.
+        //
+        // Rather than spar extensively with Diesel over these details, just
```

Comment:
This is a good call. I generally have only come out bruised and bloodied from such battles.